File: cplus/leetcode/202512/algo3742.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 网格图 DP - 2.1 基础
class Solution {
public:
    int maxPathScore(vector<vector<int>>& grid, int k) {
        // 递归 (python) 修改为递推 
        int m = grid.size(), n = grid[0].size();

        vector dp(m+1, vector(n+1, vector<int>(k+2, INT_MIN)));
        ranges::fill(dp[0][1].begin() + 1, dp[0][1].end(), 0);    // 初始化

        for (int i=0; i<m; ++i) for (int j=0; j<n; ++j) for (int K=0; K<=k; ++K){
            int new_K = K - ( grid[i][j]> 0);
            dp[i+1][j+1][K+1] = grid[i][j] + max(dp[i][j+1][new_K+1], dp[i+1][j][new_K+1]);
        }

        int ans = dp[m][n][k+1];
        return ans >= 0? ans: -1;
    }
};

File: python/algo/202512/algo3742.py
from cmath import inf
from functools import cache
from typing import List

# 动态规划 - 网格图 DP - 2.1 基础
class Solution:
    def maxPathScore(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])

        @cache
        def dfs(i: int, j: int, c: int) -> int:
            if c<0 or i < 0 or j < 0: return -inf
            if i == 0 and j == 0: return grid[0][0]

            return grid[i][j] + max(dfs(i-1, j, c-(1 if grid[i][j] else 0)), dfs(i, j-1, c-(1 if grid[i][j] else 0)))

        ans = dfs(m-1, n-1, k)
        return ans if ans >= 0 else -1
